torch/jit.py

```diff
@@ -1,26 +1,375 @@
+        else:
+            return inner
+    if callable(arg):
+        return _compile(arg)
+    else:
+        return _compile
+
+
+def trace(arg=None, nderivs=0, params=tuple()):
+    """
+    Instrument a function or module for tracing, wrapping it in a
+    :class:`TracedModule`, whose forward accepts the same arguments as the
+    original function/module, but returns a tuple consisting of the
+    *trace* of an execution, as well as the original return value.
+
+    Tracing is guaranteed not to change the semantics of the function/module
+    that is traced.
+
+    Arguments:
+        arg (optional, torch.nn.Module or function): the function or module
+            to be traced.  If `None`, `trace` returns a decorator which can be
+            applied to the function or module you want to trace.
+        nderivs (int, default 0): the number of derivatives to trace.
+            Traces of derivatives are recorded into the same trace returned
+            after executing the `forward` of the resulting module, but
+            are not present until you run `backward()` (an appropriate
+            number of times) on the resulting model.
+        params (tuple of torch.nn.Parameter): extra parameters for a traced
+            function, which do not occur as arguments to the function in
+            question.  You generally do not need this for tracing modules, as
+            the parameters of a module are automatically computed.
+
+    Example: Trace as function decorator.
+
+        >>> @jit.trace
+        >>> def f(x);
+        >>>     return x * 2
+        >>> trace, out = f(Variable(torch.randn(1)))
+
+    Example: Trace as higher order function. (Notice that trace is a *curried*
+    function; you first apply it with the function/model to trace, and then
+    apply the result with the arguments.)
+
+        >>> traced_model = jit.trace(nn.LSTMCell())
+        >>> trace, out = traced_model(input, hidden)
+
+    Example: Trace the backwards pass as function decorator.
+
+        >>> @jit.trace(nderivs=1)
+        >>> def f(x);
+        >>>     return x * 2
+        >>> trace, out = f(Variable(torch.randn(1)))
+        >>> out.sum().backward()
+        >>> print(trace)
+
+    Example: Trace the backwards pass as higher order function.
+
+        >>> traced_model = jit.trace(nn.LSTMCell(), nderivs=1)
+        >>> trace, out = traced_model(input, hidden)
+        >>> out.sum().backward()
+        >>> print(trace)
+
+    Example: Trace a function with extra parameters. (If you trace
+    a Module, parameters are automatically computed via `state_dict`).
+
+        >>> lstm = nn.LSTMCell(10, 20)
+        >>> @jit.trace(params=lstm.parameters())
+        >>> def f(a, b):
+        >>>     return lstm(a, b)
+    """
+    # TODO: handle decorating a class (not a callable)
+    def _trace(inner):
+        return TracedModule(inner, nderivs=nderivs, params=params)
+    if callable(arg):
+        return _trace(arg)
+    else:
+        return _trace
```

Comment:
This is all great for testing, but why would anyone want to permanently decorate a function such that it always return a trace? It seems like a one-time thing, so maybe it's better to expose it with a signature similar to `record_trace` we used to have

Reply:
I did it this way for consistency. If `compile` is going to be curried, then I want `trace` to be too.

I considered making it possible to say `torch.jit.trace(model, arg1, arg2)`, like how `record_trace` used to work (this wouldn't be offered for `compile`, because you basically never actually want to do that). I think we can do this, even in Python 2, but it comes at the cost of removing the explicit keyword arguments from `trace`, so you have `def trace(arg=None, *args, **kwargs)`. Then, `args is None` means decorator, `args is not None and not args` means HOF, and `args is not None and args` means apply everything immediately. I've always thought this particular mode of use is confusing, but we can put it on if people want it.

FWIW, the only reason a user would ever use `trace` is exactly because you want to do some testing/debugging. So it's not unreasonable that you would write a function/module *specifically* for the purpose of tracing it, and then delete the trace decorator when you're done.